### ISSUE TITLE
'GE' and 'NWF' paymenttypes are hard coded (Error opening order at Morris)

### ISSUE THREAD

**Issue by kentbower** [kbower/tickettest1#2131]
_Imported from trac ticket 2131.
Created by **sshah**
Opened in trac: Feb 7, 2013, 4:29:49 PM
Last modified in trac: Apr 15, 2013, 4:52:37 PM_
***
`AttributeError: Either 'Sale' object has no attribute 'balancedue' or evaluating 'balancedue' caused an AttributeError`
___
Scott,

When I try to open order 0206300AE12, I get the following error. This also
happens to be the order that I financed and funded yesterday.  I don’t know
if that has anything to do with the error or not.

I also get this error when I try to open the customer account (68) as well.

Darlene

[[Image(untitled-part.jpe)]]

***
* sshah attached [untitled-part.jpe](https://trac.retailarchitects.com/trac/attachment/ticket/2131/untitled-part.jpe) on Feb 7, 2013, 4:29:49 PM

**Comment by kentbower** [comment]
_**sshah** trac comment on Feb 7, 2013, 4:34:16 PM:_

(Changeset [[3666]](../commit/7f6923e15581c30f7c1d1b8ff0df1db8202a4a53))
* Fixes #2131 financed order used to crash when opening order screen or customer screen once they are closed.

**Comment by kentbower** [comment]
_**sshah** trac status change on Feb 7, 2013, 4:34:16 PM:_ **closed**

**Comment by kentbower** [comment]
_**kbower** trac status change on Feb 8, 2013, 7:05:50 AM:_ **reopened**

**Comment by kentbower** [comment]
_**kbower** trac comment on Feb 8, 2013, 7:06:15 AM:_

(Changeset [[3668]](../commit/7dfdd6cb1d586664e33fb2f3dd7d7d32f4cb54a5))
* revert [r3666](../commit/7f6923e15581c30f7c1d1b8ff0df1db8202a4a53), as we desire to intentionally crash when data corruption at this level.  There is a foreign key relationship there and the data needs to be fixed rather than proceed blindly.  See #2131

**Comment by kentbower** [comment]
_**kbower** trac comment on Feb 8, 2013, 7:08:19 AM:_

Fix is to correct data

**Comment by kentbower** [comment]
_**kbower** trac comment on Feb 8, 2013, 8:00:05 AM:_

They need to add 'GE' and 'PD' payment methods back to their MOP table.  They need to do this on LIVE as well as TRAIN (if not now, we will need to go through this exercise again in a few months in LIVE... would save much time to do it now.):


```
  1* select count(*), '[' || paymenttypeid || ']' from artrans where paymenttypeid is not null and not exists (select 0 from paymenttypes where paymenttypes.paymenttypeid = artrans.paymenttypeid) group by paymenttypeid
SQL> /

  COUNT(*) '['||
---------- -----
         2 [GE]
        50 [PD]

  1* select count(*), '[' || paymenttypeid || ']' from artransarchive artrans  where paymenttypeid is not null and not exists (select 0 from paymenttypes where paymenttypes.paymenttypeid = artrans.paymenttypeid) group by paymenttypeid
SQL> /

  COUNT(*) '['||
---------- -----
       106 [PD]
```

**Comment by kentbower** [comment]
_**kbower** trac status change on Feb 8, 2013, 8:00:05 AM:_ **closed**

**Comment by kentbower** [comment]
_**kbower** trac comment on Feb 8, 2013, 8:06:22 AM:_

Crap, we aren't hard coding 'GE' as a payment method are we!?

**Comment by kentbower** [comment]
_**kbower** trac status change on Feb 8, 2013, 8:06:22 AM:_ **reopened**

**Comment by kentbower** [comment]
_**kbower** trac comment on Feb 8, 2013, 8:22:28 AM:_

In addition to `'GE'`, this also needs to be fixed, it's El Dorado specific and can't be in Loft standard:
```
tran.paymenttypeid = u'NWF'
```


Also, there is a fk for mediumtypeid CONSTRAINT "ASPTRN_ASPMED_FKD" FOREIGN KEY ("MEDIUM_TP_CD") REFERENCES "AR"."ASP_MEDIUM" ("MEDIUM_TP_CD"), which means these also cannot be hardcoded:

```
fintran.mediumtypeid = u'GE'
```
and
```
fintran.mediumtypeid = u'NOR'
```

Again, those are specific to whatever retailer first implemented and cannot stay hard coded in Loft.

**Comment by kentbower** [comment]
_**kbower** trac comment on Feb 14, 2013, 6:43:32 AM:_

(Changeset [[3692]](../commit/f1f875b5f41c0887c9d63512594867ea7d0f40af))
* try to hit any potential errors before third party financing communications, see #2131 and [#2144](2144) also

**Comment by kentbower** [comment]
_**kbower** trac comment on Feb 14, 2013, 7:13:36 AM:_

Please also take the time to unite these processes for third party communication into a single framework with controller/order.py base class and sub classes such that the base class can deal with all the common code... 90% of the code is copy paste same code that is hard to maintain together.

They aren't behaviorally consistent with regards to FinanceCompanyTran(), either, which they should be.  One of them only inserts on success and one inserts on communication and updates on success.  Both should do the latter.

Example:
```python
# in order.py
class ThirdPartyFunding(object):

   def __init___(self):
       self.permission = self.api_token + u'funding'
       self.paymenttypeparam = self.api_token + u'_funding_paymenttype'

   def fundorders(self, orders):


        # ... all 
        # ... the common code
        # ... goes in the framework of the superclass


        if not orders:
            raise APIError("You must pass a list of orders in order to perform funding on them.")
    
        response = ResponseObject()
        response.results = []
        error_orders = u''
        financecompanyid = u''
        if not Permission.user_has(self.permission):
            response.errors = True
            response.notification = ['{PermissionsError}']
            response.notification_msg = Permission.err_msg(self.permission)
        elif not SystemParameter.value(self.paymenttypeparam):
            response.errors = True
            response.notification = ['{SystemParameterError}']
            response.notification_msg = ['{PaymentTypeID}', [[self.displayname]]]
        else:
            orders = utils.decodejson(orders)
            self.financecompany = DBSession.query(FinanceCompany).filter_by(interfaceid=self.api_token).one()
            financecompanyid = self.financecompany.financecompanyid
            # User object attributes will be expired after transaction.commit() so we will detached it now.
            self.user = request.identity['user']
            DBSession.expunge(self.user)
            DBSession.expunge(self.financecompany)
            batchid = DBSession.execute(metadata.sequences['fintransactions_batch_id_seq'])
            for order in orders:
                # REFACTOR: technically financecompanyid needs to be checked *after* it is selected for update below
                if order['status'] != u'F' or order['type'] != u'SAL' or order['financecompanyid'] != financecompanyid:
                    response.errors = True
                    error_orders += '\n' + order['orderid'] + ' Invalid Status'
                else:
                    try:
                        DBSession.query(FinanceCompanyTran).filter_by(orderid=order['orderid'],status=u'S').one()
                    except orm_exc.NoResultFound:
                        ord = DBSession.query(Order).for_update_of(*Order.default_forupdate).get(order['orderid'])
                        if SystemParameter.value(u'transfer_liability_at_close') == u'Y':
                            if not ord.financecompany.customerid:
                                response.errors = True
                                error_orders += '\n' + order['orderid'] + ': Finance Company has no customerid.'
                                continue
       ....
       ....


class GeFunding(ThirdPartyFunding):

   api_token = u'ge_money'
   displayname = u'GE Money'

   def setup_soap_tran(self, order):
       #... specific soap setup for GE

   def send_request(self):
       # return success boolean and response


class WellsFargoFunding(ThirdPartyFunding):

   api_token = u'wells_fargo'
   displayname = u'Wells Fargo'

   def setup_soap_tran(self, order):
       #... specific soap setup for wells fargo

   def send_request(self):
       # return success boolean and response

```

The actual controller methods should look like this:
```python
    @expose_ie_json
    @modifies_database
    def gemoney_fund_orders(self, orders=None):
        return GeFunding().fundorders(orders)

    @expose_ie_json
    @modifies_database
    def wells_fargo_fund_orders(self, orders=None):
        return WellsFargoFunding().fundorders(orders)
```

**Comment by kentbower** [comment]
_**sjohnson** trac comment on Feb 14, 2013, 10:29:35 AM:_

Darlene,
There is one more table called AR.ASP_TRN where a record is inserted when the funding happens.  There is a field called MEDIUM_TP_CD.  Below, we can see that we are populating it wrong right now.  Can you find out what the actual value should be?  It appears that it should be 'GEC' but where do we get it from?  We cannot hard code GEC. 

SQL> select count(*),MEDIUM_TP_CD from  AR.ASP_TRN group by MEDIUM_TP_CD;
COUNT(*) MED
25117.....AMX
3...........GE
142647...GEC
158113...VI
22832....DIS
114927...MC
6 rows selected.
___
I think it lives in the table **ASP**.
Darlene

**Comment by kentbower** [comment]
_**kbower** trac comment on Feb 14, 2013, 10:45:14 AM:_

[[3697]](../commit/7d9f04750acc5e353938673412463f2cc3641446)

**Comment by kentbower** [comment]
_**sshah** trac comment on Feb 14, 2013, 10:49:42 AM:_

(Changeset [[3698]](../commit/e5b9ecad84026f43caeec2c7355468c13c367a31))
* Refs #2131

**Comment by kentbower** [comment]
_**kbower** trac comment on Feb 14, 2013, 11:48:22 AM:_

(Changeset [[3699]](../commit/478387b02b9971eeb1a1614fa414612c4540307b))
* re #2131 datasetup for thirdparty financing El Dorado and Kanes

**Comment by kentbower** [comment]
_**sshah** trac comment on Apr 9, 2013, 4:09:09 PM:_

(Changeset [[4012]](../commit/c84735fbdffed2cd87332bfe881d7adb6a036477))
* Fixes #2131 and [#2239](2239). Setup interface with GE Ashley. Make all the financing company interface to work with common class which will be easy to expand.

**Comment by kentbower** [comment]
_**sshah** trac status change on Apr 9, 2013, 4:09:09 PM:_ **closed**

**Comment by kentbower** [comment]
_**kbower** trac comment on Apr 15, 2013, 4:16:57 PM:_

`AttributeError: type object 'CreditApplication' has no attribute 'submission_response'`

Can no longer load kanes rules.  `submission_response` method was removed, but rules weren't searched.

**Comment by kentbower** [comment]
_**kbower** trac status change on Apr 15, 2013, 4:16:57 PM:_ **reopened**

**Comment by kentbower** [comment]
_**sshah** trac comment on Apr 15, 2013, 4:52:37 PM:_

(Changeset [[4028]](../commit/367b5ec55d000a60adf6e5c341508c68a0abf640))
* Fixes #2131 Ability to override credit app response based upon status.

**Comment by kentbower** [comment]
_**sshah** trac status change on Apr 15, 2013, 4:52:37 PM:_ **closed**
